docs/configuration.md

```diff
@@ -49,16 +49,28 @@ This part of the configuration concerns anything that can affect the whole site.
 
 ## Plugins
 
-You can think of Quartz plugins as a series of transformations over content.
+You can think of Quartz plugins as a series of transformations over content. Many are built-in, but community members have also made their own plugins!
 
 ![[quartz transform pipeline.png]]
 
 ```ts
+// System plugins, quartz.config.ts
 plugins: {
   transformers: [...],
   filters: [...],
   emitters: [...],
 }
+
+// Community plugins, quartz.plugins.ts
+communityPlugins: {
+  available: [
+    {
+      url: ...,
+      enabled: true/false,
+      cfg: {...},
+    }
+  ]
+}
```

Comment:
1: I don't want to do that because then suddenly a normal user has to know the difference between filters, emitters, and transformers. Unifying it into one list means anyone can inspect their own user-facing config file and properly set up something cool that they saw on GitHub or the like. Happy to revise schemas for the various objects if you think of something more idiomatic or in line with Quartz otherwise!

2: that's why it's optional.

Reply:
Also remember that the json is for devs; normal users only need to touch the ts inside quartz.plugins.ts.

Reply:
> 1: I don't want to do that because then suddenly a normal user has to know the difference between filters, emitters, and transformers. Unifying it into one list means anyone can inspect their own user-facing config file and properly set up something cool that they saw on GitHub or the like. Happy to revise schemas for the various objects if you think of something more idiomatic or in line with Quartz otherwise!

I think normal users must know about this architecture in Quartz. The schema you provided is the same; they have to know that it is a `transformers` plugin or an `emitter` plugin. So, I don't really see the reason why it is in JSON.

One of my philosophies towards configuration is that imperative often offers a better user experience compared to declarative. It induces less overhead for users to learn new concepts. Since people already tweaked the ts file, we might as well put it there.

I think we can meet in the middle here for sth like this

```ts
export const Plugins = {
	"transformers": [
		{"url/repo",
		 ...config_field
		}
	],
	"emitter": [...],
	"component": [...]
}
```

Another thing is that order matters here. Especially if transformers plugins need to run before gfm, ofm, or after?. How would we handle running third-party plugins here?

That's why I think it is better to put it in the previous config, because they will know that this will be run before let, say, latex or gfm plugins.

> 2: that's why it's optional.

oh I thought there was a logic you have. I'm good with that.

> Also remember that the json is for devs; normal users only need to touch the ts inside quartz.plugins.ts.

Exactly my point, I'm more comfortable dealing with ts files rather than JSON. I think JSON is just not a good format 😄

Reply:
> One of my philosophies towards configuration is that imperative often offers a better user experience compared to declarative.

I'm completely the opposite to you on that. I think declarative abstractions can give you a much better idea of what you're looking at if you've never touched code before. If you run a command to install a plugin you found on the internet and there's suddenly all this talk about transformers and everything is in separate files and there's all this imperative *code* (imperative means steps, code) in your file, you're not going to know what to do. With the current list, anyone can go into the file, see the brackets, and find exactly what they need to and no more. Earlier iterations had me expanding community plugins into the lists in `quartz.config.ts`, but I moved it to `build.ts` because I just didn't want the user seeing these spooky variables included by other files that if you take out certain features will break in their front facing configuration file. *Technically* that approach would have been imperative.

Reply:
I think we have to agree to disagree here. I don't find any plugin systems that you can just "copy and paste," and it "just works."

If you are already thinking about plugins, chances are you might look up the documentation about the architecture of the framework.

I personally have not found yet a single use case of declarative configuration that is more efficient than imperative language other than Kubernetes, but we are not building a distributed system to run applications in production, innit?